### ISSUE TITLE
Options to increase batch size? (or to reduce number of kernel launches?)

### ISSUE THREAD

**Issue by hughperkins** [karpathy/char-rnn#55]
This is more of a question than an issue, so feel free to redirect me eg to torch newsgroup or... wherever is most convenient.  Just wondering what options could be possible (if any) to somehow increase the quantity of data we are sending to the gpu each iteration?  Currently, it seems that, using the default settings, there are several thousand kernel launches a second, which is a lot.  I dont have any good ideas for how to reduce this, just asking the question really :-)  (Hmmm... random idea... maybe can somehow 'unroll' some of the loops, so we do one, or just a few, kernel launches, rather than thousands?).

On nvidia, the kernel launches seem not to be a major issue.  On amd, I notice that the kernel launches are the main rate-limiting step.  Eg changing batch_size from 50 to 1024 leaves the iteration time entirely unchanged, so an entire epoch just takes 10 seconds, but on the other hand, almost no learning takes place during those 10 seconds, so that seems like not an ideal approach?

**Comment by karpathy** [comment]
interesting! What's being sent to the GPU is very small amount of information: just the indices X and Y, so it's odd that that would be the bottleneck. What exactly causes a kernel launch?

It's possible to ship the whole array of training indices X,Y to the GPU (this probably isnt a lot of storage), inside the Data Provider class, perhaps. Then everything would be on GPU.

**Comment by hughperkins** [comment]
A kernel is just a c99 program that runs on the gpu, or actually can be c++ in CUDA.  Kernel launch means starting a gpu program once.  Each time one calls 'forward' or 'backward' or a torch.nn module, one runs at least one program on the gpu.  In char-rnn, the network contains 40 torch.nn modules, so it has to start gpu programs 40 times for forwards, and another 40 times for backwards, total 80 times.  Within an iteration we have to do this seq_length times, total 80 * 50 = 4000 gpu program starts - kernel launches - per iteration.  On nvidia gpus, the latency for starting a program once appears to be small.  On other gpus, it can be around 100microseconds or so per launch, so 4000 * 100 microseconds = 0.4 seconds, just for starting the gpu programs.  Not including the time to actually run them, or shift data around.

**Comment by hughperkins** [comment]
What I might do is monkey-path nngraph, at runtime, so that it will join some of the network modules together, transparently.  For example, I think a bunch of those nn.Narrow modules could be eliminated, automatically, which might remove 10-20 launches per iteration, ie 10-20% performance improvement.  Anyway, I will take a look at doing this, and see how possible this might be.

tldr; : blah blah blah maybe faster blah blah blah :-)

**Comment by hughperkins** [comment]
By the way, here is a picture of the network modules for one single seq_length iteration.  It has to traverse this 50 times per iteration :-P  http://deepcl.hughperkins.com/lstm.fg.png
